Release/1.0.0

KochTobi opened the pull request:
# Changelog

This project adheres to [Semantic Versioning](https://semver.org/).

## 1.0.0 (2022-01-28)

* Add first working implementation to report updates based on a point in time.
* The application fails when any problems arise retrieving samples

KochTobi committed:
end of first session

sven1103 committed:
Provide main class and basic sample tracking service

sven1103 committed:
Code clean up

sven1103 committed:
Commit unsuccessfull service auth

sven1103 committed:
Provide location service

KochTobi committed:
Merge pull request #2 from qbicsoftware/feature/location-service

Location service for a given user id

sven1103 committed:
Introduce last update search service (#5)

Adds a last update search service, that reads the last
update search date from a textfile on the file system.

The interface also provides a method to save a new
Instant from the latest performed sample search.

sven1103 committed:
Add dependency for configuration processor

sven1103 committed:
Add openbis connector (#4)

Provides an openBIS LIMS query service implementation,
that enables the app to query the last updated samples.

Co-authored-by: wow-such-code <andreas.friedrich@uni-tuebingen.de>
Co-authored-by: Tobias Koch <KochTobi@users.noreply.github.com>

KochTobi committed:
Add CHANGELOG.md

KochTobi committed:
Implement update sink (#6)

* Add DtoMapper

* Add database config

* implement services

* implement put request

* Create Person entity

* Make it optional

* Use hibernate

* User correct object for db property assignment

Co-authored-by: Sven Fillinger <sven.fillinger@qbic.uni-tuebingen.de>

KochTobi committed:
DM-164 Orchestrate application  (#8)

* Write successful execution time for the first-time run.

* Add command line parser and command

* Implement logic

* Add tests

* Fix Sample construction in RealLimsQueryService

* Remove `println`

* Make the last update search time point optional

In case no previous update was done this should be empty.

* Update the README.md

* Ensure only successful runs where work is done are remembered

Previously to this commit runs with the `-h` or the `-V` arguments were remembered as the last update time point

KochTobi committed:
Update README.md

Fix `Run the app` for newer spring boot versions

KochTobi committed:
Create .github.settings.xml

KochTobi committed:
Add distribution management section

KochTobi committed:
Add site for report generation

KochTobi committed:
Update generate-reports.yml

Java version 1.11 -> 1.17

KochTobi committed:
Fix issues with the development branch (#10)

* Remove unused configuration property

* Use correct method to getPerson

* Use `person` table instead of `persons`

* Fix Typo

* Remove JSON mappings for location without status

* Use correct json property name for location status

* Fix samples endpoint

* Make Result throw an exception for illegal state

* Add arrival time to sample updates

* Take care to log errors as well.

* Move error handling up so that there a exception is thrown before processing.

* Update Spec description

* Update CHANGELOG.md

KochTobi committed:
Merge branch 'main' into release/1.0.0

KochTobi committed:
Update CHANGELOG.md

KochTobi committed:
Merge remote-tracking branch 'origin/release/1.0.0' into release/1.0.0

Steffengreiner commented on pom.xml:
```diff
@@ -1,79 +1,259 @@
+  <artifactId>spring-minimal-template</artifactId>
+  <version>0.1.0-SNAPSHOT</version>
+  <name>spring-minimal-template</name>
+  <description>Demo project for Spring Boot</description>
```
```suggestion
  <artifactId>sample-status-reporter</artifactId>
  <version>1.0.0</version>
  <name>sample-status-reporter</name>
  <description>Spring Boot based Sample Status Reporter application</description>
```

KochTobi replied:
Good catch!

Steffengreiner commented on README.md:
```diff
@@ -3,32 +3,39 @@
-# Requirements
-
-To run this app, you need to have version 17 of a **Java JRE** or **JDK** installed (e.g. **Zulu**).
+## System setup
+For this application to be run the following environment variables need to be set:
+
+| Name                                | Description                                                            |
+|-------------------------------------|------------------------------------------------------------------------|
+| `LAST_UPDATE_FILE`                  | A path to a persistent file. The last successful run is stored here.   |
+| `LIMS_PASSWORD`                     | The password to access the OpenBiS LIMS                                |
+| `LIMS_SERVER_URL`                   | The URL to the OpenBiS LIMS API                                        |
+| `LIMS_USER`                         | The user to access the OpenBiS LIMS                                    |
+| `SAMPLE_TRACKING_AUTH_PASSWORD`     | The password for the sample tracking user                              |
+| `SAMPLE_TRACKING_AUTH_USER`         | The username for the sample tracking service                           |
```
Nit: Maybe the Readmes can be merged to also include the description and listing in main?

KochTobi replied:
I think this is completely outdated information relating to the spring example application. Am I wrong?

Steffengreiner replied:
Ah you're right. 👁️

Steffengreiner commented on pom.xml:
```diff
@@ -1,79 +1,259 @@
+      <artifactId>spring-web</artifactId>
+    </dependency>
+    <!-- dubious openBIS dependency, nobody knows the necessity of it, but the openBIS API client
+    has it as dependency. We also have no license information. We no that this is bad practise,
```
NIT 
```suggestion
    has it as dependency. We also have no license information. We know that this is bad practise,
```

Steffengreiner commented on pom.xml:
```diff
@@ -1,79 +1,259 @@
+    </dependency>
+    <!-- dubious openBIS dependency, nobody knows the necessity of it, but the openBIS API client
+    has it as dependency. We also have no license information. We no that this is bad practise,
+    and we are ashamed.However, pressure is high to deliver the milestone, and we cannot clean
+    up the mess others have done at the moment.
```
Nit: I understand the need to explain oneself but i think it's fine to just mark this as "has to be investigated in the future". There's no need to be ashamed ;) 
```suggestion
```

KochTobi replied:
Hihi, I think we can remove this yes.

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/ReporterApp.groovy:
```diff
@@ -0,0 +1,50 @@
+ *
+ * <p>The entry-point of the reporter application.</p>
+ *
+ * @since 0.1.0
```
Nit: I'd argue that since only makes sense for the actually released versions
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/Result.groovy:
```diff
@@ -0,0 +1,156 @@
+ * </pre>
+ * @param <V> the value of this result in case it is OK
+ * @param <E> the type of error this result can hold
+ * @since 0.1.0
```
Nit: 
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/QbicSampleStatusReporter.groovy:
```diff
@@ -0,0 +1,40 @@
+    @Autowired
+    private LocationService locationService
+
+
+
```
Nit: Include groovydocs for @param and method description

KochTobi replied:
This is handled by the interface method AFAIK

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/Sample.groovy:
```diff
@@ -0,0 +1,23 @@
+/**
+ * <b>Value Class Sample</b>
+ *
+ * @since 0.1.0
```
Nit: 
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/SampleStatusProvider.groovy:
```diff
@@ -0,0 +1,16 @@
+import java.time.Instant
+
+/**
+ * <b><short description></b>
+ *
+ * <p><detailed description></p>
+ *
+ * @since <version tag>
```
Missing groovydocs

KochTobi replied:
this class will be removed (was replaed by the lims connector). good catch

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/SampleStatusReporter.groovy:
```diff
@@ -0,0 +1,14 @@
+ *
+ * <p><detailed description></p>
+ *
+ * @since <version tag>
```
Missing groovydocs

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/SampleUpdate.groovy:
```diff
@@ -0,0 +1,22 @@
+/**
+ * <b>Value Class SampleUpdate</b>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/SampleUpdate.groovy:
```diff
@@ -0,0 +1,22 @@
+@EqualsAndHashCode
+class SampleUpdate {
+    Sample sample
+    String updatedStatus
```
Question: Do you think it would make sense to change this from String to a dedicated Enum holding all possible values. (I guess it's dependent on how the statuses are stored in the LIMS)

KochTobi replied:
No by doing that we would depend on data-model-lib again and would need adjustment anytime something changes. I think it is better to let the request to the service fail for unknown statuses and validate the status there.

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/LimsQueryService.groovy:
```diff
@@ -0,0 +1,38 @@
+ * <p>Provides access to a laboratory information system and enables clients to submit simple
+ * sample queries.</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/LimsQueryService.groovy:
```diff
@@ -0,0 +1,38 @@
+     * @param updatedSince an instance that configures the query from which time-point in the past
+     * updated samples shall be listed.
+     * @return a list of updated samples
+     * @since 0.1.0
```
```suggestion
     * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/Location.groovy:
```diff
@@ -0,0 +1,22 @@
+ * <p><detailed description></p>
+ *
+ * @since <version tag>
+ */
```
Missing groovydocs

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/LocationService.groovy:
```diff
@@ -0,0 +1,23 @@
+ *
+ * <p>Provides the current location of the configured LIMS.</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/LocationService.groovy:
```diff
@@ -0,0 +1,23 @@
+     * @return an {@link Optional} wrapping a matching current {@link Location} or empty, if none was found.
+     * @throws ServiceException in case the service cannot retrieve the location due to for example connection issues.
+     * Must not be thrown, if the internal query was successful but no matching location was found. Instead, return an {@link Optional#empty}.
+     * @since 0.1.0
```
```suggestion
     * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/Person.groovy:
```diff
@@ -0,0 +1,106 @@
+/**
+ * <p>A users object that represents a users responsible for a location</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/Person.groovy:
```diff
@@ -0,0 +1,106 @@
+import javax.persistence.*
+
+/**
+ * <p>A users object that represents a users responsible for a location</p>
```
Nit: 
```suggestion
 * <p>A users object that represents a user responsible for a location</p>
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/Person.groovy:
```diff
@@ -0,0 +1,106 @@
+@ToString
+@Entity
+@Table(name = "person")
+class Person {
```
This is some magic!  🧙

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy:
```diff
@@ -0,0 +1,35 @@
+     * @param location the new location with a sample status already set
+     * @param status sample status to be set.
+     * @throws SampleUpdateException in case the sample update was unsuccessful
+     * @since 1.0.0
```
Missing @params for Person and Instant

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy:
```diff
@@ -0,0 +1,35 @@
+     * @return the location that this user is associated with
+     * @since 1.0.0
+     */
+     Optional<Location> getLocationForUser(String userId)
```
```suggestion
     Optional<Location> getLocationForUser(String userId) throws SampleUpdateException
```

KochTobi replied:
Does it throw this exception? I do not think so

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/ServiceException.groovy:
```diff
@@ -0,0 +1,19 @@
+ * <p>Shall be used, when unexpected exceptions occur during a service task execution, to indicate
+ * to the client, that some action might be advised here.</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/UpdateSearchService.groovy:
```diff
@@ -0,0 +1,29 @@
+ *
+ * <p>Provides the time point of the last update search</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/UpdateSearchService.groovy:
```diff
@@ -0,0 +1,29 @@
+     * <p>Returns the last update search time point.</p>
+     * @return the last update search time point.
+     *
+     * @since 0.1.0
```
```suggestion
     * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/api/UpdateSearchService.groovy:
```diff
@@ -0,0 +1,29 @@
+     * <p>Saves the last search time-point persistently.</p>
+     * @param lastSearchTimePoint the time-point of the last search
+     *
+     * @since 0.1.0
```
```suggestion
     * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/commands/ReportSinceInstant.groovy:
```diff
@@ -0,0 +1,93 @@
+   * @see java.lang.Thread#run()
+   */
+  @Override
+  void run() {
```
```suggestion
  void run() throws RuntimeException {
```

KochTobi replied:
I do not think we need to declare that. The RuntimeException is only thrown if something goes wrong, which should not be expected.

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/LocationService.groovy:
```diff
@@ -0,0 +1,26 @@
+ *
+ * <p><detailed description></p>
+ *
+ * @since <version tag>
```
Missing Groovydocs

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/NcctLocationService.groovy:
```diff
@@ -0,0 +1,43 @@
+ * <p><detailed description></p>
+ *
+ * @since <version tag>
+ */
```
Missing groovydocs

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/QbicSampleTrackingService.groovy:
```diff
@@ -0,0 +1,208 @@
+ *
+ * <More detailed description - When to use, what it solves, etc.>
+ *
+ * @since <version tag>
```
Missing groovydocs

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/QbicSampleTrackingService.groovy:
```diff
@@ -0,0 +1,208 @@
+        }
+    }
+
+    private static class DtoMapper {
```
Nit: Add groovydocs explaining mapping process

KochTobi replied:
I do not think we need this as this is simple copy and pasting of values around into objects.

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/RealLimsQueryService.groovy:
```diff
@@ -0,0 +1,130 @@
+ * Enables the client to request sample information from the connected LIMS.
+ * </p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/RealLimsQueryService.groovy:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    /**
+     * <b>Class AuthenticationException</b>
+     *
+     * <p>Small authentication exception class that can be used to indicate authentication exceptions</p>
+     */
+    class AuthenticationException extends RuntimeException {
+
+        AuthenticationException(String message) {
+            super(message)
+        }
+    }
+}
```
Nit: Could be extracted into a seperate class

KochTobi replied:
This is a separate class. As it is only used here, I do not see any need to move it to a different file

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/utils/SampleStatusMapper.groovy:
```diff
@@ -0,0 +1,63 @@
+ *
+ * <p>Takes a String value and tries to map it to a known sample status.</p>
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/main/groovy/life/qbic/samplestatus/reporter/services/utils/SampleStatusMapper.groovy:
```diff
@@ -0,0 +1,63 @@
+        return result
+    }
+
+    /**
+     * <b>Class MappingException</b>
+     * <p>Small mapping exception class that can be used when sample status mapping fails</p>
+     */
+    class MappingException extends RuntimeException {
+
+        MappingException(String message) {
+            super(message)
+        }
+    }
+}
```
Nit: Could be extracted into a seperate class and maybe reused for the DTOmapper class in `QbicSampleTrackingService`

KochTobi replied:
I think it is better to have different exceptions as this indicates different origins/meanings.

Steffengreiner commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
-messages.file=messages.txt
+# Properties for the sample tracking service implementation
```
Should there be default values here? I'd argue that either the values can be retrieved from the set environment variables or the reporter should stop

KochTobi replied:
You can see the default values after the `:`. Those are taken whenever no environment variable is set. Default values make sense as they also contribute to understanding what is expected here.

KochTobi replied:
The app will fail with the default values anyhow

Steffengreiner commented on src/main/resources/log4j2.xml:
```diff
@@ -0,0 +1,34 @@
+    </Console>
+
+    <RollingFile name="RollingFile"
+      fileName="./logs/samplestatus-reporter.log"
```
Nit: Consistent Naming schema
```suggestion
      fileName="./logs/sample-status-reporter.log"
```

Steffengreiner commented on src/main/resources/log4j2.xml:
```diff
@@ -0,0 +1,34 @@
+
+    <RollingFile name="RollingFile"
+      fileName="./logs/samplestatus-reporter.log"
+      filePattern="./logs/$${date:yyyy-MM}/samplestatus-reporter-%d{-dd-MMMM-yyyy}-%i.log.gz">
```
Nit: Consistent Naming schema
```suggestion
      filePattern="./logs/$${date:yyyy-MM}/sample-status-reporter-%d{-dd-MMMM-yyyy}-%i.log.gz">
```

Steffengreiner commented on src/test/groovy/life.qbic.samplestatus.reporter/ResultSpec.groovy:
```diff
@@ -0,0 +1,166 @@
+/**
+ * Tests for the {@link Result} class.
+ *
+ * @since 0.1.0
```
```suggestion
 * @since 1.0.0
```

Steffengreiner commented on src/test/groovy/life.qbic.samplestatus.reporter/ResultSpec.groovy:
```diff
@@ -0,0 +1,166 @@
+        stringResult.getError().getMessage() == exceptionMessage
+    }
+
+    def "Mapping a function to a result with value must the target result type with target value type and target value"() {
```
```suggestion
    def "Mapping a function to a result with value must return the target result type with target value type and target value"() {
```

Steffengreiner commented on src/test/groovy/life.qbic.samplestatus.reporter/ResultSpec.groovy:
```diff
@@ -0,0 +1,166 @@
+        lengthResult.getValue() == message.length() // 6
+    }
+
+    def "Mapping a function to a result with error must the target result type with target value type and input error"() {
```
```suggestion
    def "Mapping a function to a result with error must return the target result type with target value type and input error"() {
```

Steffengreiner reviewed (changes requested):
Impressive Stuff @KochTobi !  
I nitpicked mostly missing or outdated documentation. 
I'm also curious why the application needs to define new objects for DTOs from the data-model-lib such Address, Affiliation, Sample, Location.

KochTobi commented:
> for DTOs from the data-model-lib such Address, Affiliation, Sample, Location.

Because they are bad and only introduce problems 😄 This application uses hibernate to communicate with our database directly, eliminating the need for implementation of CRUD controllers.

KochTobi committed:
Update pom.xml

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
Correct typo

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
Update pom.xml

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
Update since tag

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
update since tag

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
Clean up code

Improve Spec wording

Change samplestatus to sample-status

Add JavaDoc

Update JavaDoc

Update since tag

Remove extra documentation

Remove unused interfaces

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

KochTobi committed:
Format code with google-java-format 1.13.0

Steffengreiner reviewed (approved):
LGTM 🚀
